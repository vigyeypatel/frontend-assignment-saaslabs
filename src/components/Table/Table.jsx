import React from "react";
import TableRow from "../TableRow/TableRow";
import './Table.style.css'

/**
 * Table component for displaying data.
 * @param {Array} columns - Headers of table
 * @param {Array} data - data to display.
 */
const Table = ({ columns ,data }) => {
  return (
    <div className="tableContainer">
    <table className={"table"}>
      <thead>
        <tr>
            {columns.map((column) => (
              <th key={column.toString()}>{column}</th>
            ))}
          </tr>
      </thead>
      <tbody>
        {data.length > 0 ? (
          data.map((item) => <TableRow key={item.toString()} item={item} columns={columns} />)
        ) : (
          <tr>
            <td colSpan="4">No Data Available</td>
          </tr>
        )}
      </tbody>
    </table>
    </div>
  );
};

export default React.memo(Table);
