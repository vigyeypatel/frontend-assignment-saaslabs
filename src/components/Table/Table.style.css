  /* Main container for the table */
.tableContainer {
    margin: 20px auto;
    padding: 20px;
    max-width: 90%;
    border: 2px solid #ddd;
    border-radius: 8px;
    background-color: #f9f9f9;
    box-shadow: 0 4px 10px rgba(0, 0, 0, 0.1);
    animation: fadeIn 1s ease-out;
    overflow-x: auto; /* Ensure responsiveness for smaller screens */
  }
  
  /* Table styling */
  .table {
    width: 100%;
    border-collapse: collapse;
    font-family: Arial, sans-serif;
    font-size: 16px;
    color: #333;
  }
  
  /* Table header styling */
  .table thead th {
    background-color: #007bff;
    color: white;
    text-align: left;
    padding: 10px 15px;
    text-transform: capitalize;
    border: 1px solid #007bff;
    font-weight: bold;
    transition: background-color 0.3s ease;
  }
  
  /* Hover effect for table headers */
  .table thead th:hover {
    background-color: #0056b3;
    cursor: pointer;
  }
  
  /* Table body styling */
  .table tbody tr {
    transition: background-color 0.3s ease;
  }
  
  .table tbody tr:nth-child(even) {
    background-color: #f2f2f2;
  }
  
  .table tbody tr:nth-child(odd) {
    background-color: #ffffff;
  }
  
  /* Hover effect for rows */
  .table tbody tr:hover {
    background-color: #e8f4ff; /* Subtle blue highlight */
  }
  
  /* Table cell styling */
  .table tbody td {
    padding: 10px 15px;
    border: 1px solid #ddd;
    text-align: left;
  }
  
  /* 'No Data Available' row styling */
  .table tbody tr td[colspan] {
    text-align: center;
    font-size: 18px;
    color: #666;
    padding: 20px;
    font-style: italic;
    animation: pulse 1.5s infinite ease-in-out;
  }
  
  /* Animation for 'No Data Available' */
  @keyframes pulse {
    0% {
      color: #666;
    }
    50% {
      color: #999;
    }
    100% {
      color: #666;
    }
  }
  
  /* Responsive table for small screens */
  @media (max-width: 768px) {
    .tableContainer {
      padding: 10px;
    }
  
    .table thead th {
      font-size: 14px;
      padding: 8px 10px;
    }
  
    .table tbody td {
      font-size: 14px;
      padding: 8px 10px;
    }
  
    .table tbody tr td[colspan] {
      font-size: 16px;
    }
  }
  
  /* Animation for the container fade-in */
  @keyframes fadeIn {
    from {
      opacity: 0;
      transform: translateY(20px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }
  