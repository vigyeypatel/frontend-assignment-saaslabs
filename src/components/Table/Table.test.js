import React from "react";
import { render, screen } from "@testing-library/react";
import Table from "./Table";

jest.mock("../TableRow/TableRow", () => {
  return ({ item, columns }) => (
    <tr data-testid="table-row">
      {columns.map((col, idx) => (
        <td key={idx}>{item[col]}</td>
      ))}
    </tr>
  );
});

describe("Table Component", () => {
  const columns = ["Date", "Amount", "Category", "Description"];
  const data = [
    {
      Date: "2023-12-25",
      Amount: "$100",
      Category: "Groceries",
      Description: "Christmas shopping",
    },
    {
      Date: "2023-12-24",
      Amount: "$50",
      Category: "Dining",
      Description: "Dinner with friends",
    },
  ];

  test("renders table with correct headers", () => {
    render(<Table columns={columns} data={data} />);
    columns.forEach((column) => {
      expect(screen.getByText(column)).toBeInTheDocument();
    });
  });

  test("displays table rows correctly", () => {
    render(<Table columns={columns} data={data} />);
    const rows = screen.getAllByTestId("table-row");
    expect(rows).toHaveLength(data.length);
  });

  test("displays 'No Data Available' when data is empty", () => {
    render(<Table columns={columns} data={[]} />);
    expect(screen.getByText("No Data Available")).toBeInTheDocument();
  });

  test("passes data to TableRow component", () => {
    render(<Table columns={columns} data={data} />);
    data.forEach((item) => {
      columns.forEach((col) => {
        expect(screen.getByText(item[col])).toBeInTheDocument();
      });
    });
  });

  test("handles missing columns or data gracefully", () => {
    const { container } = render(<Table columns={[]} data={[]} />);
    expect(container.querySelector("thead")).toBeInTheDocument();
    expect(screen.getByText("No Data Available")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<Table columns={columns} data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
