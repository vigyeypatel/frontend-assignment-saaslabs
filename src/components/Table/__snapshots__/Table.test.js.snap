// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="tableContainer"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Date
          </th>
          <th>
            Amount
          </th>
          <th>
            Category
          </th>
          <th>
            Description
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="table-row"
        >
          <td>
            2023-12-25
          </td>
          <td>
            $100
          </td>
          <td>
            Groceries
          </td>
          <td>
            Christmas shopping
          </td>
        </tr>
        <tr
          data-testid="table-row"
        >
          <td>
            2023-12-24
          </td>
          <td>
            $50
          </td>
          <td>
            Dining
          </td>
          <td>
            Dinner with friends
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
